lib/translator/write.cc

```diff
@@ -0,0 +1,161 @@
+#include "write.h"
+
+#include "common.h"
+
+#include "absl/types/span.h"
+
+namespace translator {
+
+// anonymous namespace for helper functions and variables
+namespace {
+
+StatusCode BuildPRInfo(uint8_t wrprotect, uint8_t& pr_info) {
+  uint8_t pract, prchk;
+
+  switch (wrprotect) {
+    case 0b000:
+      pract = 1;
+      prchk = 0b000;
+      break;
+    case (0b001 || 0b101):
+      pract = 0;
+      prchk = 0b111;
+      break;
+    case 0b010:
+      pract = 0;
+      prchk = 0b011;
+      break;
+    case 0b011:
+      pract = 0;
+      prchk = 0b00;
+      break;
+    case 0b100:
+      pract = 0;
+      prchk = 0b100;
+      break;
+    default:
+      // All other codes shall result in command termination with CHECK
+      // CONDITION status, ILLEGAL REQUEST sense key, and ILLEGAL FIELD IN CDB
+      // additional sense code.
+      DebugLog("Invalid WriteProtect Code for PRInfo.");
+      return StatusCode::kFailure;
+      break;
+  }
+
+  pr_info |= prchk | (pract << 3);
+  return StatusCode::kSuccess;
+}
+
+// This function is populates the nvme::GenericQueueEntryCmd object with fields
+// that are common to all Write Commands (6, 10, 12, 16) Refer to Section 5.7
+// (https://nvmexpress.org/wp-content/uploads/NVM_Express_-_SCSI_Translation_Reference-1_5_20150624_Gold.pdf)
+StatusCode LegacyWrite(uint64_t lba, nvme::GenericQueueEntryCmd& nvme_cmd,
+                       Allocation& allocation) {
+  StatusCode status_code = allocation.SetPages(1, 1);
+
+  if (status_code != StatusCode::kSuccess) {
+    return status_code;
+  }
+
+  nvme_cmd = {
+      .opc = static_cast<uint8_t>(nvme::NvmOpcode::kWrite),
+      .psdt = 0,
+      .mptr = allocation.mdata_addr,
+  };
+
+  nvme_cmd.dptr.prp.prp1 = allocation.data_addr;
+
+  // TODO: convert to Little-Endian as lba is in Big-Endian because it is taken
+  // from SCSI Write struct.
+  nvme_cmd.cdw[0] |= lba;
+  nvme_cmd.cdw[1] |= (lba >> 32);
+
+  return status_code;
+}
+
+StatusCode Write(bool fua, uint8_t wrprotect, uint64_t lba,
+                 uint32_t transfer_length, nvme::GenericQueueEntryCmd& nvme_cmd,
+                 Allocation& allocation) {
+  StatusCode status_code = LegacyWrite(lba, nvme_cmd, allocation);
+
+  if (status_code != StatusCode::kSuccess) {
+    return status_code;
+  }
+
+  uint8_t pr_info;
+  status_code = BuildPRInfo(wrprotect, pr_info);
+
+  if (status_code != StatusCode::kSuccess) {
+    return status_code;
+  }
+  nvme_cmd.cdw[2] |= (transfer_length | pr_info << 26 | fua << 30);
```

Comment:
Might be worth creating a new structure that implements all the commands specific d-words of NVMe write commands so that this code is more readable and less error prone.